Docs: Add career documentation (#29)
GeekHunter-Brasil/hello-world: career/expectations/software-development-analyst/README.md

- Asks for help whenever they are feeling blocked, not productive or doesn't know what to do.
- Usually codes "by example" (e.g. copying code), and **does not understand how that code works**.
- Requires more attention, since usually gets blocked often.
- Is not proactive with new tasks. Usually needs someone else to tell them what to do.

### Hard Skills

- **Knows at least a bit of the tech stack that we use** (Ruby on Rails, React, Next.js, TypeScript, Python, GraphQL and others). Will eventually have blocks and doubts on how to develop with those technologies.
- Knows at least a bit about automated tests, and at least understands their importance. Is capable of writing a test with the help of another developer.
- Usually has a difficult time to write code that respects our [Styleguide](https://github.com/GeekHunter-Brasil/hello-world).
- Usually can't do TDD daily, but knows what it is.
- Knows at least a bit about git, but can't follow processes such as gitflow.
- Doesn't know much about database technologies, and usually has a hard time to write a SQL query.
- Knows about Object Oriented Programming, but has difficulty implementing it daily.

> - Usually **doesn't know much about software architecture**. Has a had time following our current pattern (e.g. clean or hexagonal architecture).

Author: Nice catch

- Doesn't know how to design an application according to its needs (a monolith, micro-services, serveless, etc). **Usually tries to implement a complex model where it's not needed and ends up hurting productivity**.
- **Doesn't know how to debug code**. Usually has a difficult time trying to understand why something doesn't work.

### Soft Skills

- Is able to keep their presence in synchronous (e.g. squad rituals such as dailies, plannings, reviews; pair programmings) and asynchronous moments (e.g. discussions on Slack or Basecamp).
- Is really motivated to learn more and develop their skills.

> - Usually doesn't understand their own strong or weak points, or how to advance in their career. Needs an active leadership helping them to progress.

Reviewer: ```suggestion
- Usually doesn't understand their own strengths and weaknesses, or how to advance in their career. Needs an active leadership helping them to progress.
```

> ## 🐤 Software Development Analyst

Author: I thought about this but decided to go with `Software Development Analyst` only, since the way we reference `Pleno` globally is just the position (think of it as a 100%).

Check https://outliers.com.br/2020/03/10/pleno-em-ingles/

> ## 🐤 Software Development Analyst

Reviewer 2: Based on the skills sets and expectations, I'd say our Engineering team is made of Software Engineers rather than Software Development Analysts. What you guys this?

https://www.computerscience.org/resources/software-developer-vs-software-engineer/

> ## 🐤 Software Development Analyst

Author: Makes sense, @joaophellip!

Let's update this

> ## 🐤 Software Development Analyst

Author: I agree with you @joaophellip, but I don't think we should make this change because of CLT standards.

Let's leave this as-is for now, we'll review later.

- Understands concepts such as data immutability, pure functions, and reactivity.
- **Understands concepts about software architecture, and knows how to follow common patterns** such as Clean Architecture, Hexagonal Architecture and others. Knows about their advantages and disadvantages.
- Understands how to design applications and is capable of choosing a good design according to the complexity or challenge (e.g. monoliths, micro-services, serveless, etc).
- **Knows how to debug code**. Is capable of debugging their own code, manipulating the memory and instructions to reproduce a specific context.
- Usually is capable of debugging external code (such as a dependency). For example, is capable of creating a fork + pull request in a dependency in case something does not work.
- Is capable of solving any problem that doesn't need R&D (Research & Development) efficiently.

### Soft Skills

- Is able to keep their presence in synchronous (e.g. squad rituals such as dailies, plannings, reviews; pair programmings) and asynchronous moments (e.g. discussions on Slack or Basecamp).
- Is really motivated to learn more and develop their skills.

> - Usually understands their own strong or weak points, and sees something they want to prioritize in their career. The leadership process here is usually easier, since this person is able to develop their skills with some level of autonomy.

Reviewer: ```suggestion
- Usually understands their own strengths and weaknesses, and sees something they want to prioritize in their career. The leadership process here is usually easier, since this person is able to develop their skills with some level of autonomy.
```

- Sometimes codes "by example" (e.g. copying code), **but knows what they are doing and is capable of learning on the fly**. Understands what is the best path to learn something they don't know already.
- Has experience in continuous delivery and understands that the most important part about building a software is to do it via small steps, always focusing on learning more about the problem itself.
- Is fully capable of learning new technologies and languages *on demand*, since it's easy to pivot their fundamental knowledge to a new scenario.
- **Is able to communicate their decisions (technical or not) clearly and without room for doubts, along with the criteria used in each decision**.
- Is proactive about starting new tasks. Usually tells other people what to do, since they have a solid vision of what our goals are at any given time (and is aligned with our PM/POs).
- Has a clear vision about the risks involved in software development, suggests tasks to mitigate these risks when necessary.

### Hard Skills

> - **Is fluent in most part of the tech stack that we use** (Ruby on Rails, React, Next.js, TypeScript, Python, GraphQL and others).

Reviewer 2: ```suggestion
- **Is fluent in most of the tech stack that we use** (Ruby on Rails, React, Next.js, TypeScript, Python, GraphQL and others).
```

- **Is fluent in TDD, understands the importance of it and propagates it to the team**.
- Is capable of writing tests that isolate application layers in a meaningful way (both on backend and frontend), adding value to our test pyramid.

> - Is capable of producing code according to our [Styleguide](https://github.com/GeekHunter-Brasil/hello-world) and improving upon it (making new suggestions and such).

Reviewer 2: ```suggestion
- Is capable of writing code according to our [Styleguide](https://github.com/GeekHunter-Brasil/hello-world) and improving upon it (making new suggestions and such).
```

> - Is fluent with common database technologies. Is capable of solving complex performance of distribution problems (e.g. optimizing a slow query by re-implementing it or changing the database structure).

Reviewer 2: ```suggestion
- Is fluent with common database technologies. Is capable of solving complex performance and/or modeling problems (e.g. optimizing a slow query by re-implementing it or changing the database structure).
```

- Has solid experience with other database technologies (e.g. non-relational databases, graph databases, etc). Knows how and when to apply those options in daily tasks.

> - Is fluent in git. Can solve just about any problem that may arise in a repo.

Reviewer 2: ```suggestion
- Is fluent in git. Can solve just about any problem that may arise in a repo and is capable of using different strategies to improve the development flow (ex: use squash and rebase to preserve a linear history)
```

> - Is fluent in Object Oriented Programming, knows the key points to it and is able to produce high-quality OOP code.

Reviewer 2: ```suggestion
- Is fluent in Object Oriented Programming, knows the key points to it and is able to write high-quality OOP code.
```

- Has solid experience with Functional Programming, knows advantages and disadvantages to each paradigm.
- Has solid experience with concepts such as data immutability, pure functions, reactivity.
- **Has solid experience in software architecture**. Is able to implement common patterns in a disorganized repository.
- Has solid experience in application design and is capable of choosing a good design according to the complexity or challenge (e.g. monoliths, micro-services, serveless, etc).

> - **Is fluent in debugging code** and can debug just about any problem that may arise, deeply understanding what is happening and what the problem is.

Reviewer: ```suggestion
- **Is fluent in debugging code** and can debug just about any problem that may arise, deeply understanding what is happening and what the problem is, be it an overflow, memory bound, cpu bound or io bound problem. Understands how distributed computing works, how networks affect running software and how to identify and fix bottlenecks.
```

- **Is fluent at debugging external code** (such as a dependency), and can easily fix it via a pull request or using an isolated version if needed.
- Is capable of solving any problem that doesn't need R&D (Research & Development) efficiently.
- Is capable of working in problems that need R&D conducting the research in an effective way, prioritizing the use of their own time in the best way.

### Soft Skills

- Is able to keep their presence in synchronous (e.g. squad rituals such as dailies, plannings, reviews; pair programmings) and asynchronous moments (e.g. discussions on Slack or Basecamp).
- Is really motivated to learn more and develop their skills.
- **Has experience in leadership positions, mentoring other people** (usually in technical aspects).

> - Understands their own strong or weak points, and sees something they want to prioritize in their career. The leadership process here is usually easier, since **this person is able to develop their skills with a high level of autonomy**.

Reviewer: ```suggestion
- Understands their own strengths and weaknesses, and sees something they want to prioritize in their career. The leadership process here is usually easier, since **this person is able to develop their skills with a high level of autonomy**.
```
